README.md
# TFdrive

_TFdrive_ is an algorithm for functional prediction of transcription factors (TF) driving differential gene expression. It uses a list of differentially expressed genes identified, e.g., by an RNA-seq or microarray analysis and returns a list of TFs ranked by the probability score. Currently _TFdrive_ library contains 173 human TFs.

## Method Details
Accurate predictions of TFs driving changes in gene expression is still a challenging task. Current methods are predominantly based on the available TF binding site data. However, the significance of a certain TF binding site often heavily depends on a cell type, experimental conditions and activity of other TFs. As a result, such methods tend to produce a lot of false positives. 
In contrast, _TFdrive_ predictions are based on an overlap of [KEGG pathways](https://www.genome.jp/kegg/pathway.html)/[GO terms](http://geneontology.org/) between differentially expressed genes (DEGs) and a TF. The algorithm takes into account relative pathway/term importance and the number of genes related to each of the overlapping pathways/terms. This data is further supplemented with the analysis of gene-TF association tables obtained from two manually curated databases, [TRANSFAC](https://amp.pharm.mssm.edu/Harmonizome/dataset/TRANSFAC+Curated+Transcription+Factor+Targets) and [ChEA](https://amp.pharm.mssm.edu/Harmonizome/dataset/CHEA+Transcription+Factor+Targets). These associations are predominantly inferred from TF binding site data. Considering that several TFs from the same family are often able to bind to the same motif, ratios between gene-TF association frequencies among DEGs vs. non-DEGs for each TF family are calculated using TRANSFAC and ChEA tables. Family-based scores appear to have more predictive value in this context than the scores calculated for individual TFs. As a next step, probability scores using previously created random forest model are calculated. Lastly, all the scores mentioned above are used to obtain the final probability score using logistic regression. Both random forest and logistic regression models were trained using the results of 443 human TF knockout/knockdown and overexpression experiments available at [ChEA3 website](https://amp.pharm.mssm.edu/chea3/). TFs, which were among DEGs, were considered to belong to the positive class. <br/>
The data from [GSE50588](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE50588) is used here to demonstrate _TFdrive_ predictive performance. This dataset contains the microarray results of 59 human transcription factors knockdowns. Differentially expressed TFs were assigned to the positive class. Although they presumably represent only a fraction of actual positives, the number of correctly identified TFs is expected to correlate with the number of all TFs involved in differential gene expression. Therefore, the results of such analysis may give a general impression regarding algorithm's predictive accuracy. Previously, similar approach to model evaluation was employed in the development of ChEA3 (see this [publication](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6602523/) for details). Note that for testing purposes KEGG pathway/GO term information attributed to differentially expressed TFs was excluded from the analysis to avoid data leakage.
<br/>
<br/>
<p float="left">
<img src="./images/test_ROC.png" width="30%" height="30%"> <img src="./images/test_PR.png" width="30%" height="30%">
</p>

The R version of the algorithm have similar predictive efficiency in this test (ROC AUC = 0.868, PR AUC = 0.861; see the Jupyter Notebook file [here](https://github.com/i-strielkov/tfdrive/blob/master/r/model/R_model.ipynb) for details). However, since it relies on a different logistic regression library (_glmnet_) and currently does not use random forest, the results can be slightly different as compared to the Python library. <br/>
In summary, although _TFdrive_ is not designed to discover new TF-gene interactions, it allows to identify major players driving differential gene expression among known TFs in new experimental data with a high level of precision.


## Dependencies
### Python
_TFdrive_ supports Python 3.6+. Usage requires [numpy, scipy, pandas](https://www.scipy.org/), [pyarrow](https://github.com/apache/arrow), [scikit-learn](https://github.com/scikit-learn/scikit-learn), and [joblib](https://github.com/joblib/joblib).
### R
_TFdrive_ work with R 3.6.1+ and uses _glmnet_ library to calculate the probability scores.


## Installation
### Python
Download _tfdrive_files.7z_ from [here](https://drive.google.com/open?id=1q4E1QA_M7Cq5ptrbfELlCaNxvT9bJ1z3) (210 MB) and unpack _tfdrive.py_ and the _tfdrive_data_ folder to your project folder.
### R
Use _devtools_ library for installation:
```r
devtools::install_github("i-strielkov/tfdrive/r")
```


## Usage
After importing `tfdrive`, call `tfpred` providing a list of gene Entrez IDs as an argument. The values may be either integers or strings.
Currently, the method works only with human genes. `tfpred` returns a Data Frame object containing a ranked list of TFs with probability scores associated with them. The higher the score, the more likely the TF is to be involved in the observed changes in gene expression. Note that TFs from the library, which are found among DEGs, are excluded from the final results.<br/>
To have a general idea of how the probability scores relate to the actual probability of TF involvement, consider their distribution in the example shown above:<br/><br/>
<img src="./images/prob_distrib.png" width="60%" height="60%"><br/><br/>
As you can see, the TFs with probability scores higher than 0.45 - 0.5 are much more likely to belong to the positive class than to the negative class. The distribution of probability scores for R library is expected to be slightly different.


## Future plans
- Increasing TF library and training dataset
- A model for prediction of mouse TFs
- Other tweaks to the algorithm


## License
This project is licensed under the MIT License - see the LICENSE.md file for details


changelog
# Changelog
All notable changes to this project will be documented in this file.


## 22-05-2020
- Corrected pathway weight calculation
- Refined set of features used for logistic regression
- Now pathway/term score is not dependent on number of genes in a query,
  which makes the results more reliable

## 19-02-2020
- R package is now available

## 17-02-2020
- The way how missing data for GO terms is imputed was corrected.

## 06-02-2020
- The number of TFs in the library has been increased from 135 to 173. 

## 29-01-2020
- First Python version of TFdrive. 

r/tests/testthat.R
library(testthat)
library(tfdrive)

test_check("tfdrive")


r/tests/testthat/test-tfpred.R
context('tfpred')

test_that('data intact', {
  load('sysdata.rda')
  expect_true(all(c("gene_id","pathway") %in% names(pathways_hsa)))
  expect_equal(nrow(pathways_hsa), 50801)
  expect_equal(names(go_hsa), c('gene_id', 'GO_term'))
  expect_equal(nrow(go_hsa), 139684)
  expect_true(all(c('gene', 'index') %in% names(tf_names)))
  expect_equal(nrow(tf_names), 174)
})

test_that("tfpred scores correct", {
  load('sysdata.rda')
  gene_ids <- c(5714, 135114,  26958,   5899, 494514,  26502, 388325,  79007,
                4688,   1428, 259307,   2537,   1117,  55096,    953,    823,
                51031,   6892,    705,  81622,   9466, 374928,  10768,   6774,
                1178,   4494,   2769, 266971,  55651,   5168,   6813,    475,
                22906,   2793,   2766,   5269,  55179,  27339, 140771,   1901,
                2882,  79064,  10184,   6844,  51701,  57827,   1687, 122509,
                4967,  23522,  85437,   9852,   9459,    581, 729230,  57407,
                10985,   7094,   7414,   7791,   2648,   9470)

  # Data frame to store results
  full_data_df <- data.frame(TF = character(),
                             TF_id = character(),
                             p_share = double(),
                             p_all = double(),
                             p_score = double(),
                             go_share = double(),
                             go_all = double(),
                             go_score = double())

  max_pw_count = max(table(pathways_hsa$pathway))
  max_go_count = max(table(go_hsa$GO_term))

  for (i in seq_len(nrow(tf_names))){
    tf_id <- tf_names[[i, 'index']]
    tf_name <- tf_names[[i, 'gene']]

    # Get unique pathways and terms related to the TF and their counts
    tf_ptws <- pathways_hsa[pathways_hsa$gene_id==tf_id,]$pathway
    tf_terms <- go_hsa[go_hsa$gene_id==tf_id,]$GO_term
    tf_pw_count <- table(tf_ptws)
    tf_term_count <- table(tf_terms)
    p_all <- length(tf_pw_count)
    go_all <- length(tf_term_count)

    # Get pathways and terms related to differentially expressed genes
    case_ptws <- pathways_hsa[pathways_hsa$gene_id %in% gene_ids,]$pathway
    case_terms <- go_hsa[go_hsa$gene_id %in% gene_ids,]$GO_term

    # Calculate relative pathway/GO term importance
    pw_df <- as.data.frame(table(case_ptws))
    pw_df['importance'] <- sapply(pw_df$Freq, function(f) dbeta(f/max_pw_count, 2, 2) )
    go_df <- as.data.frame(table(case_terms))
    go_df['importance'] <- sapply(go_df$Freq, function(f) dbeta(f/max_go_count, 2, 2) )

    # Keep only common pathways and terms
    pw_df <- pw_df[pw_df$case_ptws %in% tf_ptws,]
    go_df <- go_df[go_df$case_terms %in% tf_terms,]

    # Calculate relative pathway/term overlap and scores based on 'importance'
    p_common <- go_common <- p_score <- go_score <- p_share <- go_share <- 0.0
    if (p_all > 0 & nrow(pw_df) > 0) {
      p_common <- length(pw_df$case_ptws[pw_df$case_ptws %in% names(tf_pw_count)])
      p_score <- sum(sapply(seq_len(nrow(pw_df)), function(f)  100 * pw_df$importance[f] * pw_df$Freq[f] / length(gene_ids)))
      p_share <- p_common/p_all
    }

    if (go_all > 0 & nrow(go_df) > 0) {
      go_common <- length(go_df$case_terms[go_df$case_terms %in% names(tf_term_count)])
      go_score <- sum(sapply(seq_len(nrow(go_df)), function(f) 100 * go_df$importance[f] * go_df$Freq[f] / length(gene_ids)))
      go_share <- go_common/go_all
    }

    full_data_df <- rbind(full_data_df,
                          data.frame(TF = as.character(tf_name),
                                     TF_id = as.character(tf_id),
                                     p_share = as.double(p_share),
                                     p_all = as.double(p_all),
                                     p_score = as.double(p_score),
                                     go_share = as.double(go_share),
                                     go_all = as.double(go_all),
                                     go_score = as.double(go_score)),
                          stringsAsFactors=FALSE)
  }

  # Check whether there are enough pathways/terms associated with each TF
  full_data_df <- full_data_df[full_data_df$p_all>2,]
  full_data_df <- full_data_df[full_data_df$go_all>2,]

  # Predictions of Logistic Regression model
  lr_pred <- predict(lr_model,
                     as.matrix(full_data_df[, c('p_share',
                                                'p_score',
                                                'go_share',
                                                'go_score')]),
                     type="response")
  full_data_df$LR_prob <- as.vector(lr_pred)

  # Preparing results
  full_data_df <- full_data_df[!full_data_df$TF_id %in% gene_ids,]
  full_data_df <- full_data_df[order(full_data_df$LR_prob, decreasing = TRUE),]
  rownames(full_data_df) <- 1:nrow(full_data_df)

  expect_equal(as.character(full_data_df$TF[1]), 'STAT4')
  expect_equal(as.character(full_data_df$TF_id[1]), '6775')
  expect_equal(round(full_data_df$LR_prob[1], 6), 0.483113)
  expect_equal(round(full_data_df$p_score[1], 6), 0.565435)
  expect_equal(round(full_data_df$go_score[1], 6), 0.323160)
})


python/tfdrive.py
import numpy as _np
import pandas as _pd
import os as _os
import pyarrow as _pyarrow
import scipy.stats as _st
from collections import Counter as _Counter
from sklearn.linear_model import LogisticRegression as _LogisticRegression
from sklearn.ensemble import RandomForestClassifier as _RandomForestClassifier
from joblib import load as _load


def tfpred(gene_ids, species = 'hsa'):
    '''
    Predicts transcription factors driving differential gene expression.
    Parameters:
    ----------
    gene_ids : array_like, int
        1-D array of Entrez Gene IDs of differentially expressed genes.

    species : str
        Three-letter species ID. Currently, function works only with human genes. 
        Therefore, 'hsa' is the default value and should not be changed.

    Returns : DataFrame
        Returns a list of transcription factors ranked according to their probability 
        scores.
    '''
    # Check the input
    if type(gene_ids) is str:
        raise ValueError("'gene_ids' should be a list of IDs")

    try:
        gene_ids = list(gene_ids)
    except TypeError:
        print("'gene_ids' argument is not iterable.")
        raise

    if len(_np.array(gene_ids).shape) > 1:
        raise ValueError('The list of gene IDs should be flat')

    if len(gene_ids)<10:
        raise ValueError('The list of genes is too short')

    for i in range(len(gene_ids)):
        if type(gene_ids[i]) is not int:
            try:
                gene_ids[i] = int(gene_ids[i])
            except ValueError:
                print("An element of 'gene_ids' cannot be converted to an integer.")
                raise

    if species!='hsa':
        raise ValueError("Only 'hsa' option is currently supported.")
    
    # Load data
    file_list = ['tfdrive_data/pathways_hsa_full_hg38.parquet',
                 'tfdrive_data/go_terms_hsa_processes.parquet',
                 'tfdrive_data/tf_names_hsa.parquet',
                 'tfdrive_data/tf_families_hsa.parquet',
                 'tfdrive_data/tf_matrix_transfac.parquet',
                 'tfdrive_data/tf_matrix_chea.parquet',
                 'tfdrive_data/rf_mod.joblib',
                 'tfdrive_data/log_mod.joblib']
                 
    if not all([_os.path.isfile(f) for f in file_list]):
        raise OSError("Some required data files are missing.")

    # Table of gene ID to pathway associations
    pathways_db = _pd.read_parquet('tfdrive_data/pathways_hsa_full_hg38.parquet')
    pathways_db.drop_duplicates(['gene_id','pathway'],keep='first', inplace=True)

    gene_ids = list(set(gene_ids))
    if len([g for g in gene_ids if g in pathways_db['gene_id'].values]) < 10:
        raise ValueError("Too few or none of provided gene IDs were found in the database.")

    # Table of gene ID to GO terms (from 'processes' cathegory) associations
    go_terms_db = _pd.read_parquet('tfdrive_data/go_terms_hsa_processes.parquet')
    go_terms_db.drop_duplicates(['gene_id','GO_term'],keep='first', inplace=True)

    # Table of TF to TF ID associations
    tf_names_db = _pd.read_parquet('tfdrive_data/tf_names_hsa.parquet')

    # Load TF classification table
    tf_families_df = _pd.read_parquet('tfdrive_data/tf_families_hsa.parquet')

    # Tables of TF-gene interactions
    transfac_df = _pd.read_parquet('tfdrive_data/tf_matrix_transfac.parquet')
    chea_df = _pd.read_parquet('tfdrive_data/tf_matrix_chea.parquet')

    # Arbitrary distribution for calculating 'pathway importance'
    dist = _st.beta(2,2)

    # Prediction models
    rf_model = _load('tfdrive_data/rf_mod.joblib')
    log_model = _load('tfdrive_data/log_mod.joblib') 

    full_data_df = _pd.DataFrame(columns=['TF', 'TF_id', 'p_share', 
                                         'all_p', 'p_score', 'go_share', 
                                         'all_go', 'go_score'])

    for tf_id in tf_names_db.index:
        tf_name = tf_names_db.at[tf_id, 'gene']  

        # Get unique pathways and terms related to the TF and their counts
        tf_ptws = list(pathways_db[pathways_db['gene_id'].isin([tf_id])]['pathway'])
        tf_terms = list(go_terms_db[go_terms_db['gene_id'].isin([tf_id])]['GO_term'])
        tf_pw_count = _Counter(tf_ptws)
        tf_term_count = _Counter(tf_terms)
        all_p = len(tf_pw_count)
        all_go = len(tf_term_count)

        # Get unique pathways and terms related to differentially expressed genes
        # and their counts
        case_ptws = list(pathways_db[pathways_db['gene_id'].isin(gene_ids)]['pathway'])
        case_terms = list(go_terms_db[go_terms_db['gene_id'].isin(gene_ids)]['GO_term'])
        case_ptw_count = _Counter(case_ptws)
        case_term_count = _Counter(case_terms)

        # Calculate relative pathway/GO term importance
        max_pw_count = max(_Counter(pathways_db['pathway'].values).values())
        max_go_count = max(_Counter(go_terms_db['GO_term'].values).values())

        pw_df = _pd.DataFrame.from_dict(case_ptw_count, orient='index',columns=['count'])
        pw_df['importance'] = pw_df['count'].apply(lambda x: x/max_pw_count)
        pw_df['importance']  = dist.pdf(pw_df['importance'])

        go_df = _pd.DataFrame.from_dict(case_term_count, orient='index',columns=['count'])
        go_df['importance'] = go_df['count'].apply(lambda x: x/max_go_count)
        go_df['importance']  = dist.pdf(go_df['importance'])

        # Calculate relative pathway/term overlap and scores based on 'importance'
        common_p = 0
        common_go = 0
        p_score = 0.0
        go_score = 0.0
        coef = 100
        
        if all_p > 0:
            for i in tf_pw_count:
                if i in case_ptw_count:
                    common_p += 1
                    p_score += coef * pw_df.at[i, 'importance'] * pw_df.at[i, 'count'] / len(gene_ids)

        if all_go > 0:
            for t in tf_term_count:
                if t in case_term_count:
                    common_go += 1
                    go_score += coef * go_df.at[t, 'importance'] * go_df.at[t, 'count'] / len(gene_ids)
  
        p_share = common_p/all_p if all_p > 0 else 0.0
        go_share = common_go/all_go if all_go > 0 else 0.0
        
        # Add to DataFrame
        full_data_df.loc[len(full_data_df)] = [tf_name, tf_id, p_share, all_p,  
                                               p_score, go_share, all_go, go_score]

    #Drop TFs which have less than 3 pathways associated with them
    full_data_df = full_data_df[full_data_df['all_p']>2].copy(deep=True)
    full_data_df = full_data_df[full_data_df['all_go']>2].copy(deep=True)

    # Calculate relative interaction frequencies for each family for each group
    tf_group_scores_df = _pd.DataFrame(columns=['family', 'transfac', 'chea'])
    gene_ids_str = [str(g) for g in gene_ids]
    for f in tf_families_df.index:
        tf_ids = tf_families_df.at[f, 'ids']
        # Transfac
        t_value = 0.0
        t_tf_ids = [t for t in tf_ids if t in transfac_df.columns]
        t_g_ids = [h for h in gene_ids_str if h in transfac_df.index]    
        if (len(t_tf_ids) > 0) & (len(t_g_ids) > 0):
            t_value = (transfac_df.loc[t_g_ids, t_tf_ids].sum().sum() / len(t_g_ids)) / \
                      (transfac_df[t_tf_ids].sum().sum() / len(transfac_df))
        # ChEA
        c_value = 0.0
        c_tf_ids = [c for c in tf_ids if c in chea_df.columns]
        c_g_ids = [h for h in gene_ids_str if h in chea_df.index]    
        if (len(c_tf_ids) > 0) & (len(c_g_ids) > 0):
            c_value = (chea_df.loc[c_g_ids, c_tf_ids].sum().sum() / len(c_g_ids)) / \
                      (chea_df[c_tf_ids].sum().sum() / len(chea_df))
        tf_group_scores_df.loc[len(tf_group_scores_df)] = [f, t_value, c_value]
            
    # Create TF to TF family association table
    tf_family_pairs = _pd.DataFrame(columns=['family'])
    for i in tf_families_df.index:
        for tf in tf_families_df.at[i, 'ids']:
            tf_family_pairs.loc[tf] = [i]
            
    # Assign values to full_data_df
    full_data_df['transfac'] = 0.0
    full_data_df['chea'] = 0.0
    for i in full_data_df.index:
        tf_id = str(full_data_df.at[i, 'TF_id'])
        if tf_id in tf_family_pairs.index:
            family = tf_family_pairs.at[tf_id, 'family']
            full_data_df.at[i, 'transfac'] = tf_group_scores_df[tf_group_scores_df['family'] == family]['transfac'].values[0]
            full_data_df.at[i, 'chea'] = tf_group_scores_df[tf_group_scores_df['family'] == family]['chea'].values[0]
        
        # If TF is not in the TF families, calculate individual values
        else:
            t_value = 0.0
            t_g_ids = [h for h in gene_ids_str if h in transfac_df.index]    
            if (tf_id in transfac_df.columns) & (len(t_g_ids) > 0):
                t_value = (transfac_df.loc[t_g_ids, tf_id].sum().sum() / len(t_g_ids)) / \
                    (transfac_df[tf_id].sum().sum() / len(transfac_df))
                full_data_df.at[i, 'transfac'] = t_value 

            c_value = 0.0
            c_g_ids = [h for h in gene_ids_str if h in chea_df.index]    
            if (tf_id in chea_df.columns) & (len(c_g_ids) > 0):
                c_value = (chea_df.loc[c_g_ids, tf_id].sum().sum() / len(c_g_ids)) / \
                    (chea_df[tf_id].sum().sum() / len(chea_df))
                full_data_df.at[i, 'chea'] = c_value 

    # Fill zeroes if only one ('chea' or 'transfac') value is available
    for i in full_data_df.index:
        if full_data_df.at[i, 'chea'] == 0.0:
            full_data_df.at[i, 'chea'] = full_data_df.at[i, 'transfac']
        if full_data_df.at[i, 'transfac'] == 0.0:
            full_data_df.at[i, 'transfac'] = full_data_df.at[i, 'chea'] 

    # Predictions of Random Forest model
    full_data_df['RF_prob'] = rf_model.predict_proba(full_data_df[['TF_id',
                                                                    'p_share','all_p',
                                                                    'p_score', 'go_share',
                                                                    'all_go', 'go_score',
                                                                    'transfac','chea']].values)[:,[1]]

    # Predictions of Logistic Regression model
    full_data_df['LR_prob'] = log_model.predict_proba(full_data_df[['p_share','p_score', 
                                                'go_share', 'go_score', 'RF_prob']].values)[:,[1]]


    full_data_df.drop(columns=['RF_prob', 'p_share', 'all_p', 'p_score', 'go_share', 'all_go', 
                               'go_score', 'transfac', 'chea'], inplace=True)
    intersection = [g for g in tf_names_db.index if g in gene_ids]
    full_data_df = full_data_df[~full_data_df['TF_id'].isin(intersection)].copy(deep=True)
    full_data_df.sort_values('LR_prob', ascending=False, inplace=True)
    full_data_df.reset_index(drop=True, inplace=True)

    return full_data_df

python/jupyter_notebooks/pwcal_hsa.py
import numpy as np
import pandas as pd
import pyarrow
from multiprocessing import Pool
from collections import Counter
from itertools import repeat
import scipy.stats as st

def __get_ptws(ids, pathways_db, go_terms_db):
    '''Get list of patways/terms based on gene IDs'''
    pathways = list(pathways_db[pathways_db['gene_id'].isin(ids)]['pathway'])
    go_terms = list(go_terms_db[go_terms_db['gene_id'].isin(ids)]['GO_term'])
    return pathways, go_terms


def __pcal(n, tf_id, tf_name, cases_db, pathways_db, go_terms_db, dist, exclude_deg_tf):
    ''' Find an overlap between KEGG pathways/GO terms related to differentially 
    expressed genes in each case and each TF in the library and calculate corresponding
    scores based on this overlap and the relative pathway/term importance'''
    full_data_df = pd.DataFrame(columns=['case_id', 'case', 'TF', 'TF_ids', 'p_share', 
                                        'all_p', 'p_score', 'go_share', 'all_go', 
                                        'go_score'])
    indx = 0
    print(f'Processing TF {n+1}: {tf_name} ID: {tf_id}')  

    # Get unique pathways and terms related to the TF and their counts 
    tf_ptws, tf_terms = __get_ptws([tf_id], pathways_db, go_terms_db)
    tf_pw_count = Counter(tf_ptws)
    tf_term_count = Counter(tf_terms)
    all_p = len(tf_pw_count)
    all_go = len(tf_term_count)

    total_ptws_count = Counter(pathways_db['pathway'].values)
    total_term_count = Counter(go_terms_db['GO_term'].values)
    total_pw_df = pd.DataFrame.from_dict(total_ptws_count, orient='index',columns=['count'])
    total_go_df = pd.DataFrame.from_dict(total_term_count, orient='index',columns=['count'])
    max_pw_count = total_pw_df['count'].max()
    max_go_count = total_go_df['count'].max()

    for c in cases_db['case'].index: 
        # Get unique pathways and terms related to differentially expressed genes in each
        # case and their counts. If TF from the library is among the genes, its pathways
        # and terms can be excluded to avoid data leakage (for the testing purposes).
        if exclude_deg_tf:
            case_ids = [g for g in cases_db.at[c,'degs_entrez'] if g != tf_id]
        else:
            case_ids = cases_db.at[c,'degs_entrez'] 
        case_ptws, case_terms = __get_ptws(case_ids, pathways_db, go_terms_db)
        if (len(case_ptws)==0 or len(case_terms)==0): continue 
        case_ptw_count = Counter(case_ptws)
        case_term_count = Counter(case_terms)

        # Calculate relative pathway importance for this case
        # The score is calculated based on beta distribution
        pw_df = pd.DataFrame.from_dict(case_ptw_count, orient='index',columns=['count'])
        pw_df['importance'] = pw_df['count'].apply(lambda x: x/max_pw_count)
        pw_df['importance']  = dist.pdf(pw_df['importance'])

        # Calculate relative GO term importance for this case
        go_df = pd.DataFrame.from_dict(case_term_count, orient='index',columns=['count'])
        go_df['importance'] = go_df['count'].apply(lambda x: x/max_go_count)
        go_df['importance']  = dist.pdf(go_df['importance'])

        # Calculate relative pathway/term overlap and scores based on 'importance'
        common_p = 0
        common_go = 0
        p_score = 0.0
        go_score = 0.0
        coef = 100
        if all_p > 0:
            for i in tf_pw_count:
                if i in pw_df.index:
                    common_p += 1
                    p_score += coef * pw_df.at[i, 'importance'] * pw_df.at[i, 'count'] / len(case_ids)
        if all_go > 0:
            for t in tf_term_count:
                if t in go_df.index:
                    common_go +=1
                    go_score += coef * go_df.at[t, 'importance'] * go_df.at[t, 'count'] / len(case_ids)

        p_share = common_p/all_p if all_p > 0 else 0.0
        go_share = common_go/all_go if all_go > 0 else 0.0
        
        # Add to DataFrame
        full_data_df.loc[indx] = [c, cases_db.at[c,'case'], tf_name, tf_id, p_share, all_p,  
                                  p_score, go_share, all_go, go_score]
        indx +=1

    return full_data_df
        

def pwcal (cases_db, exclude_deg_tf = False, n_jobs = 2):
    '''
    Calculate pathway and GO term similarity between TFs and DEGs
    Parameters:
    ----------
    cases_db : DataFrame object
        Preprocessed table containing lists of differentially expressed genes

    exclude_deg_tf : bool, default False
        If True, KEGG pathways and GO terms related to TFs, which appear in the list of DEGs
        differentially expressed genes, are not included.

    n_jobs : int, default 2
        Number of parallel processes
    
    Returns : DataFrame
        Returns DataFrame with corresponding pathway/term scores
    '''

    # Table of TF to TF ID associations
    tf_names_db = pd.read_parquet('data/tf_names_hsa.parquet')
    
    # Table of gene ID to pathway associations
    pathways_db = pd.read_parquet('data/pathways_hsa_full_hg38.parquet')
    pathways_db.drop_duplicates(['gene_id','pathway'],keep='first', inplace=True)
    
    # Table of gene ID to GO terms (from 'processes' cathegory) associations
    go_terms_db = pd.read_parquet('data/go_terms_hsa_processes.parquet')
    go_terms_db.drop_duplicates(['gene_id','GO_term'],keep='first', inplace=True)
    
    # Arbitrary distribution for calculating 'pathway importance'
    dist = st.beta(2,2)

    # Create a process pool 
    pool = Pool(processes=n_jobs) 
    print(" {} processes started...".format(n_jobs))

    results = pool.starmap(__pcal, zip(
        range(len(tf_names_db.index)),
        list(tf_names_db.index),
        list(tf_names_db['gene'].values),
        repeat(cases_db),
        repeat(pathways_db),
        repeat(go_terms_db),
        repeat(dist),
        repeat(exclude_deg_tf)
        ))

    pool.close()
    pool.join()
    results = pd.concat(results)
    results.reset_index(drop=True, inplace=True)
    print('Finished')
    return results





r/R/tfpred.r
#' Predicts transcription factors driving differential gene expression based on a list of
#' differentially expressed genes
#'
#' @param gene_ids vector of Entrez Gene IDs of differentially expressed genes
#' @param species three-letter species ID (currently, function works only with human genes ('hsa'))
#'
#' @return data frame with transcription factors ranked according to their probability scores
#'
#' @examples
#' tfpred(c(5714, 135114, 26958, 5899, 494514, 26502, 388325, 79007, 4688, 1428, 259307))
#'
#' @import glmnet
#'
#' @export
tfpred <- function(gene_ids, species = 'hsa'){

  #Checks the input data
  gene_ids <- unlist(gene_ids)
  gene_ids <- unique(gene_ids)
  if (!is.null(dim(gene_ids))) stop('The list of gene IDs should be flat')
  if (length(gene_ids) < 10) stop('The list of genes is too short')
  gene_ids <- as.integer(gene_ids)
  if (any(is.na(gene_ids))) stop("'gene_ids' should be a list of numeric Entrez IDs")
  if (species!='hsa') stop("Only 'hsa' option is currently supported.")

  # Used data
  # pathways_hsa - table of pathways attributed to human genes
  # go_hsa - table of GO terms attributed to human genes
  # tf_names - list of TFs
  # lr_model - logistic regression model

  if (nrow(pathways_hsa[pathways_hsa$gene_id %in% gene_ids,]) < 10) {
    stop ("Too few or none of provided gene IDs were found in the database.")
  }

  # Data frame to store results
  full_data_df <- data.frame(TF = character(),
                             TF_id = character(),
                             p_share = double(),
                             p_all = double(),
                             p_score = double(),
                             go_share = double(),
                             go_all = double(),
                             go_score = double())

  max_pw_count = max(table(pathways_hsa$pathway))
  max_go_count = max(table(go_hsa$GO_term))

  for (i in seq_len(nrow(tf_names))){
    tf_id <- tf_names[[i, 'index']]
    tf_name <- tf_names[[i, 'gene']]

    # Get unique pathways and terms related to the TF and their counts
    tf_ptws <- pathways_hsa[pathways_hsa$gene_id==tf_id,]$pathway
    tf_terms <- go_hsa[go_hsa$gene_id==tf_id,]$GO_term
    tf_pw_count <- table(tf_ptws)
    tf_term_count <- table(tf_terms)
    p_all <- length(tf_pw_count)
    go_all <- length(tf_term_count)

    # Get pathways and terms related to differentially expressed genes
    case_ptws <- pathways_hsa[pathways_hsa$gene_id %in% gene_ids,]$pathway
    case_terms <- go_hsa[go_hsa$gene_id %in% gene_ids,]$GO_term

    # Calculate relative pathway/GO term importance
    pw_df <- as.data.frame(table(case_ptws))
    pw_df['importance'] <- sapply(pw_df$Freq, function(f) dbeta(f/max_pw_count, 2, 2) )
    go_df <- as.data.frame(table(case_terms))
    go_df['importance'] <- sapply(go_df$Freq, function(f) dbeta(f/max_go_count, 2, 2) )

    # Keep only common pathways and terms
    pw_df <- pw_df[pw_df$case_ptws %in% tf_ptws,]
    go_df <- go_df[go_df$case_terms %in% tf_terms,]

    # Calculate relative pathway/term overlap and scores based on 'importance'
    p_common <- go_common <- p_score <- go_score <- p_share <- go_share <- 0.0
    if (p_all > 0 & nrow(pw_df) > 0) {
      p_common <- length(pw_df$case_ptws[pw_df$case_ptws %in% names(tf_pw_count)])
      p_score <- sum(sapply(seq_len(nrow(pw_df)), function(f)  100 * pw_df$importance[f] * pw_df$Freq[f] / length(gene_ids)))
      p_share <- p_common/p_all
    }

    if (go_all > 0 & nrow(go_df) > 0) {
      go_common <- length(go_df$case_terms[go_df$case_terms %in% names(tf_term_count)])
      go_score <- sum(sapply(seq_len(nrow(go_df)), function(f) 100 * go_df$importance[f] * go_df$Freq[f] / length(gene_ids)))
      go_share <- go_common/go_all
    }

    full_data_df <- rbind(full_data_df,
                          data.frame(TF = as.character(tf_name),
                                     TF_id = as.character(tf_id),
                                     p_share = as.double(p_share),
                                     p_all = as.double(p_all),
                                     p_score = as.double(p_score),
                                     go_share = as.double(go_share),
                                     go_all = as.double(go_all),
                                     go_score = as.double(go_score)),
                          stringsAsFactors=FALSE)
  }

  # Check whether there are enough pathways/terms associated with each TF
  full_data_df <- full_data_df[full_data_df$p_all>2,]
  full_data_df <- full_data_df[full_data_df$go_all>2,]

  # Predictions of Logistic Regression model
  lr_pred <- predict(lr_model,
                     as.matrix(full_data_df[, c('p_share',
                                                'p_score',
                                                'go_share',
                                                'go_score')]),
                     type="response")
  full_data_df$LR_prob <- as.vector(lr_pred)

  # Preparing results
  full_data_df <- full_data_df[,c('TF', 'TF_id', 'LR_prob')]
  full_data_df <- full_data_df[!full_data_df$TF_id %in% gene_ids,]
  full_data_df <- full_data_df[order(full_data_df$LR_prob, decreasing = TRUE),]
  rownames(full_data_df) <- 1:nrow(full_data_df)

  full_data_df
}
